Loop/Managers/WatchDataManager.swift

```diff
@@ -334,6 +372,52 @@ extension WatchDataManager {
 
         return items.joined(separator: "\n")
     }
+    
+    private func getSleepStartTime(start: Date, end: Date? = nil, sampleLimit: Int = 30, _ completion: @escaping (_ result: SleepStoreResult<Date>) -> Void) {
+        let predicate = HKQuery.predicateForSamples(withStart: start, end: end, options: [])
+        
+        getSleepStartTime(matching: predicate, sampleLimit: sampleLimit, completion)
+    }
+    
+    private func getSleepStartTime(matching predicate: NSPredicate, sampleLimit: Int, _ completion: @escaping (_ result: SleepStoreResult<Date>) -> Void) {
+        let sleepType = HKObjectType.categoryType(forIdentifier: HKCategoryTypeIdentifier.sleepAnalysis)!
+        // get more-recent values first
+        let sortDescriptor = NSSortDescriptor(key: HKSampleSortIdentifierEndDate, ascending: false)
+        
+        let query = HKSampleQuery(sampleType: sleepType, predicate: predicate, limit: sampleLimit, sortDescriptors: [sortDescriptor]) { (query, samples, error) in
+
+            if let error = error {
+                completion(.failure(error))
+            } else if let samples = samples as? [HKCategorySample] {
+                // find the average hour and minute components from the sleep start times
+                let average = samples.reduce(0, {$0 + $1.startDate.secondsFromMidnight()}) / samples.count
+                let averageHour = average / 3600
+                let averageMinute = average % 3600 / 60
+                
+                // find the next time that the user will go to bed, based on the averages we've computed
+                if let time = Calendar.current.nextDate(after: Date(), matching: DateComponents(hour: averageHour, minute: averageMinute), matchingPolicy: .nextTime) {
+                    completion(.success(time))
+                } else {
+                    completion(.failure(NSError()))
+                }
+            } else {
+                assertionFailure("Unknown return configuration from query \(query)")
```

Comment:
Since assertion failures trigger only in debug modes, it's best practice to still call the completion handler on this path—not that I'd expect it to happen!